### ISSUE TITLE
Remote storage queue's concurrent sends cause out-of-order samples

### ISSUE THREAD

**Issue by juliusv** [prometheus/prometheus#1843]
The remote storage queue manager fires off concurrent requests to send sample batches to the remote storage: https://github.com/prometheus/prometheus/blob/e980913cd69a85d9644ffd0d0cfeb26633e9f983/storage/remote/queue_manager.go#L183-L204

This causes samples to be sent out-of-order, which the receiving side might not be equipped to handle. Not sure what's going to happen with this interface in general, but if it's generally going to stay, we should consider removing the parallelism (or make it configurable in case people have a backend that doesn't care?).

**Comment by brian-brazil** [comment]
I think the other end will need to be able to deal with this. I'm expecting parallelism to be required to have the required throughput.

**Comment by tomwilkie** [comment]
We could have both parallelism and in order delivery.  For instance, we could shard by metric name (or fingerprint...) and then deliver in parallel.

**Comment by brian-brazil** [comment]
To get to the point where you could consider not having parallelism would require major design first (as in we'd need to completely flesh out this interface), so I think that's very much out of scope for the medium term anyway.
